Feature/delete image

rhdqors opened the pull request:
## 작업 내용
- 이미지 삭제 기능 구현

## 변경 사항
- 추가적으로 서비스단 커스텀 에러 파일 새로 생성 ErrorCode enum, BusinessException
- 서비스단 커스텀 처리 위한 globalexception 파일내 메서드 생성 

## 테스트 방법
- 컨트롤러에 user 인증부분 주석이나 삭제하고 터미널에 아래 값 입력
- Invoke-RestMethod -Method DELETE -Uri http://localhost:8080/api/v1/products/1/images/9994

rhdqors committed:
refactor: 조회 기능 구현 위한 cqrs 적용

rhdqors committed:
이미지 목록 조회 구현

기존 upload 구조 변경 및 조회용 query파일 생성.

rhdqors committed:
이미지 삭제 기능 구현

추가적으로 서비스단 커스텀에러 파일 생성

rhdqors committed:
Merge pull request #15 from JiHongKim15/feature/select-all-image

Feature/select all image

rhdqors committed:
이미지 목록조회 컨트롤러 리턴타입 변경

rhdqors committed:
Merge pull request #18 from JiHongKim15/feature/select-all-image

refactor: 이미지 목록조회 컨트롤러 리턴타입 변경

rhdqors committed:
Merge branch 'feature/delete-image' of https://github.com/JiHongKim15/music-sale-platform-be into feature/select-all-image

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageController.java:
```diff
@@ -51,6 +51,29 @@ public ResponseEntity<ApiResponse<List<ImageOutput>>> uploadImages(
+     * @return 삭제 완료 응답
+     */
+    @DeleteMapping("/api/v1/products/{productId}/images/{imageId}")
+    public ResponseEntity<ApiResponse<Void>> deleteImage(
```
ResponseEntity로 한번 더 안감싸도 됐던 것 같아용

rhdqors replied:
수정했습니다.

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageController.java:
```diff
@@ -51,6 +51,29 @@ public ResponseEntity<ApiResponse<List<ImageOutput>>> uploadImages(
+    public ResponseEntity<ApiResponse<Void>> deleteImage(
+            @PathVariable Long productId,
+            @PathVariable Long imageId,
+            @AuthenticationPrincipal UserPrincipal loginUser) {
```
AuthenticationPrincipal
이 어노테이션은 무엇일까요..? 
참고로 User는 나중에 session에 저장된 user 정보를 가지고 오는 방향으로 할게요!

rhdqors replied:
스프링 시큐리티에 등록된 유저 정보를 자동으로 가져오는 어노테이션으로 사용했습니다. 추후에 스프링시큐리티 사용 예정이라고 들었던 것 같아서요! 

시큐리티는 계획이 없을까요ㅕ?

JiHongKim15 replied:
시큐리티를 하긴 할건데 아직 정의된게 없어서..! 가지고 온다고 하더라도 앞에서 받는 값은 아닐 것 같아서 다음에 적용하시졍!

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageController.java:
```diff
@@ -51,6 +51,29 @@ public ResponseEntity<ApiResponse<List<ImageOutput>>> uploadImages(
+        // 이미지 삭제
+        imageUseCase.deleteImage(productId, imageId);
+
+        return ResponseEntity.status(HttpStatus.NO_CONTENT)
```
ApiResponse로 통일해주세요~

rhdqors replied:
수정했습니다.

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageController.java:
```diff
@@ -51,6 +51,29 @@ public ResponseEntity<ApiResponse<List<ImageOutput>>> uploadImages(
+            @AuthenticationPrincipal UserPrincipal loginUser) {
+
+        // 권한 검증 (기존 메서드 재사용)
+        checkUserAuthorization(loginUser, productId);
```
controller에 있는 checkUserAuthorization 아에 삭제하구, 나중에 해당 내용 검토하는 걸로 하시죠!

rhdqors replied:
네 컨트롤러에서 삭제했습니다.

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageQueryController.java:
```diff
@@ -0,0 +1,39 @@
+@RestController
+@RequestMapping("/api/v1/products")
+@RequiredArgsConstructor
+public class ImageQueryController {
```
ImageController가 있는데 ImageQueryController가 따로 ㅇㅆ는 이유는 무엇일까용..?! 만약 CQRS로 한다고 하면
ImageCommandController, ImageQueyrController 이렇게 있어야할 것 같아요!

rhdqors replied:
조회용 query하나랑 다른 기능용 controller로 사용했었는데 이름때문에 혼동이 있던 것 같습니다.  하나는 이름 command로 변경하겠스빈다.

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageQueryController.java:
```diff
@@ -0,0 +1,39 @@
+                ApiResponse.success(outputs, "이미지 조회 성공")
+        );
+    }
+
```
여기도 일괄적으로 ApiResponse로 변경 부탁드립니다!. 혹시 다른 컨트롤러에도 ApiResponse를 ResponseEntity가 감싸고 있다면 일괄적으로 변경부탁들여ㅛ!

rhdqors replied:
변경했습니다.

JiHongKim15 commented on music-api/src/main/kotlin/com/music/sale/common/GlobalExceptionHandler.kt:
```diff
@@ -93,4 +94,17 @@ class GlobalExceptionHandler {
     }
+
+    @ExceptionHandler(BusinessException::class)
+    fun handleBusinessException(e: BusinessException): ResponseEntity<ApiResponse<Unit>> {
```
여기둥..

rhdqors replied:
수정했습니다. 카톡에서 말씀하신 추후 필요한 내용 담는 방향이 좋을 것 같습니다.

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/image/port/outport/ImageQueryPort.java:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ * 이미지 조회용 포트 (영속성 계층이 구현)
+ */
```
해당 주석은 굳이 필요해 보이지 않네요..! 아래도 마찬가지같습니다.

rhdqors replied:
수정했습니다.

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/image/service/ImageService.java:
```diff
@@ -55,4 +57,19 @@ private ImageOutput createImageOutput(
+            // 이미지 삭제
+            imagePort.deleteById(imageId);
+        } catch (Exception e) {
+            throw new BusinessException(ErrorCode.IMAGE_DELETE_FAILED);
```
반드시 try catch로 감싸야할 이유가 있을까요?

rhdqors replied:
서비스에서 예외처리 해준다면 없어도 될 것 같습니다.  수정했습니다.

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/image/controller/ImageController.java:
```diff
@@ -51,6 +51,29 @@ public ResponseEntity<ApiResponse<List<ImageOutput>>> uploadImages(
+     * @param loginUser 로그인 사용자
+     * @return 삭제 완료 응답
+     */
+    @DeleteMapping("/api/v1/products/{productId}/images/{imageId}")
```
/api/v1/products는 공통적인 부분으로 컨트롤러에 전체 적용되는게 좀 더 좋지 않을까요?
그리고 혹시 이미지는 단건 삭제만 있을까요? 전체 이미지 삭제도 있으면 좋겠네요!

rhdqors replied:
requestMapping을 빼먹었던 것 같습니다. 추가했습니다. 이미지 삭제의 경우 당근마켓과 같이 상품, 이미지 등록할 때 이미지별로 x표시가 있어서 해당 이미지만 삭제되는 쪽으로 얘기를 했던 것 같은데 한번 더 얘기해보고 진행하겠습니다.

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/common/BusinessException.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Getter;
+
+@Getter
+public class BusinessException extends RuntimeException {
```
BizException 근데 코틀린에 있는거랑 겹치지 않나여? 아에 이관하신것일까요

rhdqors replied:
코틀린에 bizexception이 안보여서 따로 만들었습니다. 전체 찾기해도 안되던데 제가 못찾은걸까요?

JiHongKim15 replied:
엏,,없었나여? 없으면 만드는게 맞긴 한 것 같아요! 제가 누락했었을 수도 있긴 해서 .. 남겨두시죠!

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/common/ErrorCode.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum ErrorCode {
```
errorCode가 image만 관련된 것이 아니라면
ErrorCode class 안에 ErrorProductImage와 같이 enum을 선언하는 건 어떤거 같으신가요?

rhdqors replied:
errorcode안에 이미지 뿐만 아닌 다른 도메인도 다 포함시키면 좋을 것 같아서 찬성하는 부분입니다. 하지만 모든 에러를 이 파일에서 관리하면 혹시 깃 충돌은 괜찮을까에 대한 걱정? 이 있는 것 같습니다. 더 좋은 방법들이 있을까요?

팀원들이랑 얘기중인건  ddd 생각해서 각 도메인별로 에러 관리 클래스들을 두자 정도인 것 같습니다.

JiHongKim15 replied:
사실 error 파일은 xml에서 관리합니당 보통 하나로 관리하구, 외국어까지 고려해서 해용! 그리고 에러까지 클래스로 나누게 되면 너무 복잡해져버려서... 이런건 하나로 모으는게 나은 것 같습니당 깃 충돌이 생겨도 영역이 활실한 부분이라서요!

rhdqors replied:
그렇다면 말씀하신대로 추후 한 곳에서 관리하는 방향으로 가면 좋을 것 같네요

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/image/port/inport/ImageUseCase.java:
```diff
@@ -6,4 +6,6 @@
 
 public interface ImageUseCase {
     List<ImageOutput> uploadImage(List<UploadImageInput> inputs);
+    void deleteImage(Long productId, Long imageId);
```
근데 imageId가 productImage table에 들어가게 되면 사실상 imageId도 유일할텐데, productId를 꼭 받아야 할 이유가 있을까요??

rhdqors replied:
productId가 없으면 서비스 로직에서 유효성 검사할 때 imageId.getProductId()와 같은 코드들이 많아질 것 같은데 매개변수로 넣으면 한번에 사용 가능하지 않을까요

JiHongKim15 replied:
음 기능 정의 나름인 것 같긴 한데, 이미지는 반드시 어떤 product의 이미지를 삭제할 것이다 라는 내용으로 정의가 되면 필요할 것 같고, 아니면 유일한 값이라서 굳이 productId가 필요할까 싶긴 합니당..!
그거와 별개로 image는 increment로 id를 꼭 쌓아야할까여..?! 이미지가 많아지게되면 increment값이 괜찮을지 고민이되네용
한 번 생각해보시구 알려주세용..!!

rhdqors replied:
말씀하신대로 특정 상품의 이미지 삭제로 구현한 코드라 productId가 있어야한다 라는 생각입니다. 

이미지는 최소1장 필수, 최대 5장까지만 가능한 부분이라 이정도면 id가쌓여도 괜찮다 라고 예상됩니다.

rhdqors committed:
이미지 삭제 pr 반영

컨트롤러의 ResponseEntity 제거 후 apiresponse로 통일.
컨트롤러명 쿼리용, 일반용 분리.
이미지 삭제시 트라이 캐치 제거.
